portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -71,22 +84,20 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    
-    // if (messages.isEmpty()) {
-    //     messages = new ArrayList<Comment>();
-    // }  
 
     String commentMessage = getParameter(request, "comment-area");
     Entity messageEntity = new Entity("Comment");
     if (!commentMessage.isEmpty()) {
-        // messages.add(commentMessage);
         long timestamp = System.currentTimeMillis();
         messageEntity.setProperty("message", commentMessage);
         messageEntity.setProperty("timestamp", timestamp);
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-        datastore.put(messageEntity);
+        Key key = datastore.put(messageEntity);
```

Comment:
Thanks for the explanation. I understood the problem now. This is an interesting question. It seems the either datastore.put(messageEntity) call is completed or the datastore reading part is not consistent.

I'm not very familiar with GCP's DataStore. Usually in transactional database this is not a problem. I did some google search, it seems put is not the problem. "If a user relies on implicit transaction management and the installed policy creates a transaction, that transaction will be committed (in the case of a success) or rolled back (in the case of a failure) before the operation returns to the user " - https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/DatastoreService

I guess the Read could be the problem. Maybe you can try this option: Specify ReadPolicy.Consistency in the reading configurations in DatastoreServiceConfig.

Getting a key in the put function is strange, because you are only creating a local variable to store the return value, it doesn't affect the implementation of put function itself, thus it shouldn't make any difference. Nor to mention compiler might even optimise out the unused key local variable. Using a list in memory is usually not a good practice, because it is hard to sync the list in memory when you have more than one servers.

Reply:
Ah okay. I am also thinking about this situation right now, how to react with other users' comments. Previously I am thinking about every time the datastore changes, how would it tell the client side. So your method is just to use javascript to pull data from the datastore regularly, like every 10 seconds and don't need to care about whether the datastore has changed, right? Yah this should work, I will figure out how to use this method. Thanks Kai!

Reply:
Pulling is the only solution I'm afraid. Datastore is being replaced by Firestore, and this page (https://cloud.google.com/datastore/docs/firestore-or-datastore#in_native_mode) says there're a few improvements in Firestore over DataStore, especially **Real-time updates**

Reply:
Okay noted. So for now I will try to write the javascript background pull solution first and will go through how to use the Firestore to replace the datastore when having time. Anyway thanks very much for your suggestion as I am stuck with this problem for quite a long time.